initial PR for SDIOBlockdevice implementation

JojoS62 opened the pull request:
This is the component that was not merged into mbed-os.

JojoS62 committed:
initial commit

add SDIOBlockdevice
implementation for STM32F4

JojoS62 committed:
Update README.md

added some description

JojoS62 committed:
add uitl folder, necessary for tests

JojoS62 committed:
add implementation for STM32F7

almost the same as for STM32F4, but in F7 SDIO device is renamed to SDMMC
SDMMC1 is used for DISCO_F746NG
updated readme.md for new target

JojoS62 committed:
changed copyright notice to standard mBed

JojoS62 committed:
changed copyright to mBed standard

0xc0170 commented:
@ARMmbed/mbed-os-storage Can you review?

dannybenor commented:
@0xc0170 will do after 5.11 release

screamerbg commented:
@dannybenor We need a review sooner. This is blocking new Pelion DM boards by ST. Could you please help?

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+#define SD_DBG 0       /*!< 1 - Enable debugging */
+#define SD_CMD_TRACE 0 /*!< 1 - Enable SD command tracing */
+
+#define SD_BLOCK_DEVICE_ERROR_WOULD_BLOCK -5001           /*!< operation would block */
```
shouldn't those be in H file?

JojoS62 replied:
the debug settings are used private in the cpp source, they are not relevant for other modules. I have taken this approach from the SDBlockdevice.

offirko replied:
Those return errors should be exposed by the h file (possibly for SDBlockDevice as well)
An even better suggestion is to use the errors at mbed_error.h 

https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_error.h

    UNKNOWN                    256      Unknown error
    INVALID_ARGUMENT           257      Invalid Argument
    INVALID_DATA               258      Invalid data
    INVALID_FORMAT             259      Invalid format
    INVALID_INDEX              260      Invalid Index
    ...

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+    if (BLOCK_SIZE_HC != _cardInfo.BlockSize)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_UNSUPPORTED_BLOCKSIZE;
```
_is_initialized is true at this stage, do we need to set it back to false? Deinit SD_Card?

JojoS62 replied:
yes, it is reset in Line 153. There is a reference counting because there can be multiple calls to init and only the last call to deinit must release all resources.

dannybenor commented:
blockdevice should be a component and not a feature
This needs review from HAL team

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+
+int SDIOBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    //debug_if(SD_DBG, "read Card...\r\n");
```
remove comment if not required

JojoS62 replied:
ok

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+    return status;
+}
+
+int SDIOBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
```
please give a more meaningful and debug-able name than 'b'

JojoS62 replied:
ok

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+    if (status == MSD_OK)
+    {
+        // wait until DMA finished
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
```
Does SDIO mandatory require DMA? if not maybe it should be wrapped in an abstraction layer that will simulate the 'driver' layer?

JojoS62 replied:
I tried first the polling version, but had lots of performance problems. The data fifo must be read/written very fast, otherwise you'll get over/underrun errors. Only the DMA was fast enough to avoid these problems.

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
+            if ((HAL_GetTick() - tickstart) >= MBED_CONF_SD_TIMEOUT)
```
hmm.. I didnt understand the logic here.. seems like you're measuring the time between two sequential GetTick's .
maybe tickstart should only be initialized before the loop?

JojoS62 replied:
yes, this was faulty. Moved tickstart outside while loop.

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+        // wait until DMA finished
+        while (SD_DMA_WritePending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
```
Same as above...I think tickstart should only be initialized before the loop?

JojoS62 replied:
yes.

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+        while (SD_GetCardState() != SD_TRANSFER_OK)
+        {
+            // wait until SD ready
+            wait_ms(10);
```
why is it wait 10 msec here, and 1 msec in other places?

JojoS62 replied:
I'm testing shorter timeouts now and I have added timeout checks to all loops. In this case, the card should be ready and the check is too make sure that a previous operation is not blocking.
The timeout check with high frequent HAL_Tick() calls may cause problems because there are many calls to locking/unlocking for the ticker access.

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+    return status;
+}
+
+int SDIOBlockDevice::program(const void *b, bd_addr_t addr, bd_size_t size)
```
please replace "b" with something more meaningful

JojoS62 replied:
ok

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+        wait_ms(1);
+    }
+
+    int status = SD_WriteBlocks_DMA(buffer, addr, blockCnt);
```
same remark regarding DMA, unless all SDIO must work with DMA - it should be moved to an abstraction layer API

JojoS62 replied:
same reason, DMA is neccessary

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,378 @@
+
+bool SDIOBlockDevice::isPresent(void)
+{
+    if (_cardDetect.is_connected())
```
mbed-os style requires {} for single line if conditions as well

JojoS62 replied:
ok

offirko commented on SDIOBlockDevice.h:
```diff
@@ -0,0 +1,143 @@
+     *  @param buffer   Buffer to write blocks to
+     *  @param addr     Address of block to begin reading from
+     *  @param size     Size to read in bytes, must be a multiple of read block size
+     *  @return         0 on success, negative error code on failure
```
Please try to elaborate possible error values for each API

offirko commented on SDIOBlockDevice.h:
```diff
@@ -0,0 +1,143 @@
+    SD_Cardinfo_t _cardInfo;
+    uint32_t _card_type;
+
+    PlatformMutex _mutex;
```
Can there be other SDIOBlockDevice instances or other SDIO devices that will try to use the same DMA 'driver' ?
If so .. this per-instance mutex will not protect parallel use of the DMA 'driver' is it ok?

JojoS62 replied:
In this case with STM32F4, SDIO is a singleton because there is only one SDIO in hardware. I have not checked how to avoid to instantiate two SDIOBlockdevices, that should not be possible. There are also global vars preventing the use of multiple instances.
In STM32F7, there are devices with 2 hardware units, I have used only the first one in that implementation. In F7, the name for the device is also different, that is confusing.
Another suggestion in another conversation was to add the resource names in the constructor. That would be a cleaner solution for devices with several units, but I could not finish it until now.
This blockdevice driver was not accepted for the mbed-os, it should seperate into a HAL and device layer. I agree, but when this should be done its worth to work on the better solution. The singleton approach works now for the most boards with a SDIO interface (there are only a few). For the future, a HAL SDIO is better because it may be used for RF devices also, not only SD cards.

deepikabhavnani replied:
> Can there be other SDIOBlockDevice instances or other SDIO devices 

Limiting to single instance is good idea for now.

offirko commented:
@JojoS62 - thanks for submitting this important BlockDevice implementation. I've left several questions and remarks - thanks!

JojoS62 commented:
I have prepared some modifications as requested, but I will run the tests again first. Will push the commits tomorrow.

JojoS62 committed:
code cleanup

- removed comments
- renamed short variable names

screamerbg commented:
@JojoS62 @offirko Any update on this? We have 7 ST boards using SDIO, but we're forced to use external SD (SPI) shield.

dannybenor commented:
@screamerbg this requires HAL to go into mbedOS master. Please create a preq and ad HAL team

deepikabhavnani commented:
@dannybenor - This PR was moved to external repo for the same reason. HAL changes might take time as they are not planned, but that should not stop users to use SDIO interface. 

HAL changes will be needed to integrate with Mbed OS, but in parallel this driver can be polished and used as external. Also other partners will have reference to add to SDIO changes which will be helpful in finalizing HAL API's.

dannybenor commented:
@deepikabhavnani Yes, you are right. I missed this point. LGTM

offirko commented on SDIOBlockDevice.cpp:
```diff
@@ -0,0 +1,377 @@
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
+            if ((HAL_GetTick() - tickstart) >= MBED_CONF_SD_TIMEOUT)
```
This requires a fix in order to measure Timeout correctly.

offirko commented:
@JojoS62 - thanks for making adjustments! 
I believe the Timeout issue still requires a fix.

JojoS62 committed:
fiixed timeout checking

tickstart was inside checking loop
removed wait_ms() calls
added timeout checks for all while loops

jeromecoutant commented:
Hi

Any update ?

Thx

screamerbg commented:
@dannybenor @offirko Any update on this?

JojoS62 commented:
sorry for the delay, at the moment I'm very busy with my work.
The component is working, I've applied the most fixes. Only the error codes could be improved, but the mbed errror handling has become also complex and at the moment I have no time to update this. I have used similar codes like in SDBlockDevice, so this component is usable already.
Please review and check if there are major issues that prevent using this PR.

offirko reviewed (approved):
LGTM

deepikabhavnani reviewed (approved):
@JojoS62 - Thanks for adding this

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Please review and merge

0xc0170 commented:
> @ARMmbed/mbed-os-maintainers - Please review and merge

we can review but merging is up to @ARMmbed/mbed-os-storage team I believe

0xc0170 commented on util/fsfat_test.h:
```diff
@@ -0,0 +1,74 @@
+/** @file fsfat_test.h
+ *
+ * mbed Microcontroller Library
+ * Copyright (c) 2006-2016 ARM Limited
```
this should be 2019 (new file or at least 2018-2019)

New files should contain SPDX identifier (look at platform header files).

0xc0170 reviewed (changes requested):
Small license fix, rest looks fine

deepikabhavnani commented:
>we can review but merging is up to @ARMmbed/mbed-os-storage team I believe

Sounds good, not sure why I thought only maintainers have merge permissions on this repo

0xc0170 commented:
@ARMmbed/mbed-os-storage Can we progress? there's new PR that depends also on this one

jeromecoutant commented:
Why this PR is not progressing ?

adbridge commented:
> 
> 
> Why this PR is not progressing ?

@ARMmbed/mbed-os-storage @dannybenor Could you please answer @jeromecoutant  ?

maclobdell commented:
Just curious - would this better to now have have a generic SDIOBlockDevice in the mbed-os/components/storage/blockdevice/COMPONENT_SDIO and then specific implementations under platform-specific target code at mbed-os/targets/TARGET_<VENDOR>/.../TARGET_<BOARD>/

Or alternatively, if there isn't a HAL abstraction to be used by a generic block device driver, perhaps fully target-specific SDIO blockdevice implementations could live in the target code.  

Can we move forward with one of these approaches, or perhaps another approach?

mathias-arm committed:
Fix order order of parameters

JojoS62 committed:
added get_type() function

JojoS62 committed:
fixed printf warning

jeromecoutant commented:
@MarceloSalazar @adbridge don't forget this feature :-)

adbridge commented:
@jeromecoutant we don't deal with this repository and it doesn't form part of our mbed-os releases. @MarceloSalazar @0xc0170 do you know anything about how this repo is used/controlled ?

JojoS62 commented:
It seems I missed some notifications about updates here. I will try to implement the changes soon.
@adbridge This component was not complying the rules for an integration into Mbed-os, so it was offloaded here. It would be nice if the master branch could be filled.

@irfanAfa is this issue still pending? I don't have a NUCLEO_F746ZG, so I can check it only in theorie.

irfanafa commented:
It's good for the STM32F746. Thankyou

On Thu, 13 Aug 2020, 14:03 JojoS, <notifications@github.com> wrote:

> It seems I missed some notifications about updates here. I will try to
> implement the changes soon.
> @adbridge <https://github.com/adbridge> This component was not complying
> the rules for an integration into Mbed-os, so it was offloaded here. It
> would be nice if the master branch could be filled.
>
> @irfanafa <https://github.com/irfanafa> is this issue still pending? I
> don't have a NUCLEO_F746ZG, so I can check it only in theorie.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ARMmbed/sdio-driver/pull/1#issuecomment-673434864>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKVTDIBKONA5QCQAIPNJBGTSAPI75ANCNFSM4GC6LHLA>
> .
>

JojoS62 committed:
remove path from include
mbed file structure has changed again

JojoS62 committed:
add SDMM2 for F7 with 2 infterfaces

jeromecoutant reviewed (approved):
Didn't test, but I assume @JojoS62 did !
Thx

JojoS62 commented:
in the past month, I reworked this component for the F7. I added also support for the DISCO_F769NI, which uses the 2nd  SDMMC interface. I still need to clean up experimental stuff and run the filesystem tests again.
I experienced some problems with the data caching and added some cache flush functions. The difficult thing is, that it depends on the memory what is used. Mbed uses a 'one linker script fits all', but this is difficult to keep when F7 with D-cache and DMA is used. Same problem when I used DMA2D for graphics. I can open an issue for that subject, I want to mention it only here because it can drive you nuts...

JojoS62 committed:
update for mbed-os 6

JojoS62 committed:
add H7

JojoS62 committed:
change lib type to static

JojoS62 committed:
add sleep in wait for data Todo: use event_flags

JojoS62 committed:
use MBED_ERROR in error callback

JojoS62 committed:
use mbed error macro, add sdmmc reset as in BSP

JojoS62 committed:
disable debug break

JojoS62 committed:
set delay time in read to 0 ms

JojoS62 committed:
repeat reading until no error
